build error on Ubuntu 18, Swift 5.1 : "reference to 'SHA256_CTX' is ambiguous"

I'm trying to build swift-nio-ssl on an Ubuntu 18 machine, using Swift 5.1, as part of a server based on Vapor 4 alpha (yes I know). Everything builds fine except for the following error :

`<unknown>:0: error: reference to 'SHA256_CTX' is ambiguous
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio-ssl/Sources/CNIOBoringSSL/include/CNIOBoringSSL/base.h:411:32: note: candidate found by name lookup is 'SHA256_CTX'
typedef struct sha256_state_st SHA256_CTX;
                               ^
//usr/include/openssl/sha.h:56:3: note: candidate found by name lookup is 'SHA256_CTX'
} SHA256_CTX;
  ^
Failed to reconstruct type for $sSo10SHA256_CTXaD
Original type:
(type_alias_type decl=CAPNSOpenSSL.(file).SHA256_CTX)`

Is that error expected somehow ? I have libssl 1.1 installed on this server, on which the vapor package depends.

Sure, here it is

```
warning: 'fetch' command is deprecated; use 'resolve' instead
Fetching https://github.com/vapor/vapor.git
Fetching https://github.com/vapor/crypto-kit.git
Fetching https://github.com/vapor/routing-kit.git
Fetching https://github.com/vapor/async-kit.git
Fetching https://github.com/vapor/nio-websocket-client.git
Fetching https://github.com/kylebrowning/APNSwift
Fetching https://github.com/apple/swift-nio-extras.git
Fetching https://github.com/vapor/swift-nio-http-client.git
Fetching https://github.com/apple/swift-nio-ssl.git
Fetching https://github.com/apple/swift-log.git
Fetching https://github.com/apple/swift-nio-http2.git
Fetching https://github.com/vapor/console-kit.git
Fetching https://github.com/apple/swift-nio.git
Completed resolution in 28.70s
Cloning https://github.com/vapor/console-kit.git
Resolving https://github.com/vapor/console-kit.git at 4.0.0-alpha.1
Cloning https://github.com/vapor/nio-websocket-client.git
Resolving https://github.com/vapor/nio-websocket-client.git at 1.0.0-alpha.1
Cloning https://github.com/apple/swift-nio-ssl.git
Resolving https://github.com/apple/swift-nio-ssl.git at 2.1.1
Cloning https://github.com/apple/swift-log.git
Resolving https://github.com/apple/swift-log.git at 1.1.0
Cloning https://github.com/vapor/vapor.git
Resolving https://github.com/vapor/vapor.git at 4.0.0-alpha.1.5
Cloning https://github.com/vapor/swift-nio-http-client.git
Resolving https://github.com/vapor/swift-nio-http-client.git at 0.0.0
Cloning https://github.com/kylebrowning/APNSwift
Resolving https://github.com/kylebrowning/APNSwift at 1.0.0
Cloning https://github.com/vapor/crypto-kit.git
Resolving https://github.com/vapor/crypto-kit.git at 4.0.0-alpha.1
Cloning https://github.com/apple/swift-nio-http2.git
Resolving https://github.com/apple/swift-nio-http2.git at 1.3.0
Cloning https://github.com/apple/swift-nio-extras.git
Resolving https://github.com/apple/swift-nio-extras.git at 1.1.0
Cloning https://github.com/vapor/routing-kit.git
Resolving https://github.com/vapor/routing-kit.git at 4.0.0-alpha.1
Cloning https://github.com/vapor/async-kit.git
Resolving https://github.com/vapor/async-kit.git at 1.0.0-alpha.1
Cloning https://github.com/apple/swift-nio.git
Resolving https://github.com/apple/swift-nio.git at 2.4.0
Fetching Dependencies [Done]
Building Project ...
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketClient.swift:24:31: warning: conformance of 'HTTPRequestEncoder' to protocol 'RemovableChannelHandler' was already stated in the type's module 'NIOHTTP1'
extension HTTPRequestEncoder: RemovableChannelHandler { }
                              ^
NIOHTTP1.HTTPRequestEncoder:1:20: note: 'HTTPRequestEncoder' declares conformance to protocol 'RemovableChannelHandler' here
final public class HTTPRequestEncoder : NIO.ChannelOutboundHandler, NIO.RemovableChannelHandler {
                   ^
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/nio-websocket-client/Sources/NIOWebSocketClient/WebSocketHandler.swift:29:13: warning: variable 'frame' was never mutated; consider changing to 'let' constant
        var frame = self.unwrapInboundIn(data)
        ~~~ ^
        let
<unknown>:0: error: reference to 'SHA256_CTX' is ambiguous
/srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio-ssl/Sources/CNIOBoringSSL/include/CNIOBoringSSL/base.h:411:32: note: candidate found by name lookup is 'SHA256_CTX'
typedef struct sha256_state_st SHA256_CTX;
                               ^
//usr/include/openssl/sha.h:56:3: note: candidate found by name lookup is 'SHA256_CTX'
} SHA256_CTX;
  ^
Failed to reconstruct type for $sSo10SHA256_CTXaD
Original type:
(type_alias_type decl=CAPNSOpenSSL.(file).SHA256_CTX)
Stack dump:
0.	Program arguments: /usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift -frontend -c -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSSwiftJWT/APNSwiftJWT.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSSwiftJWT/APNSwiftJWTError.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSSwiftJWT/Data+Base64URL.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftConfiguration.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftConnection.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftErrors.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftPointerTricks.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftRequest.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftRequestContext.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftRequestEncoder.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftResponse.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftResponseDecoder.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftSigner.swift -primary-file /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSwiftStreamHandler.swift -supplementary-output-file-map /tmp/supplementaryOutputs-8a97b3 -target x86_64-unknown-linux -disable-objc-interop -sdk / -I /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio/Sources/CNIOHTTPParser/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio-ssl/Sources/CNIOBoringSSLShims/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio-ssl/Sources/CNIOBoringSSL/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio/Sources/CNIOSHA1/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio/Sources/CNIOAtomics/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio/Sources/CNIODarwin/include -I /srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio/Sources/CNIOLinux/include -enable-testing -g -module-cache-path /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/ModuleCache -swift-version 5 -Onone -D SWIFT_PACKAGE -D DEBUG -color-diagnostics -enable-anonymous-context-mangled-names -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIOHTTPParser.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIOBoringSSLShims.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/checkouts/swift-nio-ssl/Sources/CNIOBoringSSL/include/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIOSHA1.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIOAtomics.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIODarwin.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/CNIOLinux.build/module.modulemap -Xcc -fmodule-map-file=/srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/CAPNSOpenSSL/module.modulemap -parse-as-library -module-name APNSwift -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSSwiftJWT/APNSwiftJWT.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSSwiftJWT/APNSwiftJWTError.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSSwiftJWT/Data+Base64URL.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftConfiguration.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftConnection.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftErrors.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftPointerTricks.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftRequest.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftRequestContext.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftRequestEncoder.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftResponse.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftResponseDecoder.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftSigner.swift.o -o /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/APNSwift.build/APNSwiftStreamHandler.swift.o -index-store-path /srv/datadisk01/www/VideoSaverServer/.build/x86_64-unknown-linux/debug/index/store -index-system-modules 
1.	While emitting IR SIL function "@$s8APNSwift0A3JWTV6sha25633_AAE6F96E24F1CD18F931B2288965D25BLL7message3NIO10ByteBufferVAI_tF".
 for 'sha256(message:)' (at /srv/datadisk01/www/VideoSaverServer/.build/checkouts/APNSwift/Sources/APNSwift/APNSSwiftJWT/APNSwiftJWT.swift:70:13)
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x4517774]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x45153ce]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x4517b88]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f975deaa890]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f975c30de97]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f975c30f801]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6a7dc3]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6a112d]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6a0f91]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6c7eb3]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6c44d9]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x6b2d55]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x5e260b]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x54e0c7]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x54e601]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x4d4ff2]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x4d0c0e]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x473b16]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f975c2f0b97]
/usr/local/swift-5.1-DEVELOPMENT-SNAPSHOT-2019-07-09-a-ubuntu18.04/usr/bin/swift[0x47377a]

Building Project [Failed]
Error: execute(code: 1)
admin@server01:~/www/VideoSaverServer$ 

```

--- a/Sources/NIOSSL/ByteBufferBIO.swift
+++ b/Sources/NIOSSL/ByteBufferBIO.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import NIO
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 
 /// The BoringSSL entry point to write to the `ByteBufferBIO`. This thunk unwraps the user data

--- a/Sources/NIOSSL/NIOSSLHandler.swift
+++ b/Sources/NIOSSL/NIOSSLHandler.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import NIO
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIOTLS
 
 /// The base class for all NIOSSL handlers. This class cannot actually be instantiated by

--- a/Sources/NIOSSL/SSLCallbacks.swift
+++ b/Sources/NIOSSL/SSLCallbacks.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 
 /// The result of an attempt to verify an X.509 certificate.

--- a/Sources/NIOSSL/SSLCertificate.swift
+++ b/Sources/NIOSSL/SSLCertificate.swift
@@ -12,8 +12,13 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+#else
 import CNIOBoringSSL
 import CNIOBoringSSLShims
+#endif
 import NIO
 
 /// A reference to a BoringSSL Certificate object (`X509 *`).
@@ -26,7 +31,11 @@ import NIO
 /// to obtain an in-memory representation of a TLS certificate from a buffer of
 /// bytes or from a file path.
 public class NIOSSLCertificate {
-    internal let ref: UnsafeMutablePointer<X509>
+    internal let _ref: UnsafeMutableRawPointer/*<X509>*/
+
+    internal var ref: UnsafeMutablePointer<X509> {
+        return self._ref.assumingMemoryBound(to: X509.self)
+    }
 
     internal enum AlternativeName {
         case dnsName([UInt8])
@@ -39,7 +48,7 @@ public class NIOSSLCertificate {
     }
 
     private init(withReference ref: UnsafeMutablePointer<X509>) {
-        self.ref = ref
+        self._ref = UnsafeMutableRawPointer(ref) // erasing the type for @_implementationOnly import CNIOBoringSSL
     }
 
     /// Create a NIOSSLCertificate from a file at a given path in either PEM or
@@ -102,7 +111,7 @@ public class NIOSSLCertificate {
     ///
     /// In general, however, this function should be avoided in favour of one of the convenience
     /// initializers, which ensure that the lifetime of the `X509` object is better-managed.
-    static public func fromUnsafePointer(takingOwnership pointer: UnsafeMutablePointer<X509>) -> NIOSSLCertificate {
+    static func fromUnsafePointer(takingOwnership pointer: UnsafeMutablePointer<X509>) -> NIOSSLCertificate {
         return NIOSSLCertificate(withReference: pointer)
     }
 

--- a/Sources/NIOSSL/SSLConnection.swift
+++ b/Sources/NIOSSL/SSLConnection.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import NIO
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 internal let SSL_MAX_RECORD_SIZE = 16 * 1024
 

--- a/Sources/NIOSSL/SSLContext.swift
+++ b/Sources/NIOSSL/SSLContext.swift
@@ -13,8 +13,13 @@
 //===----------------------------------------------------------------------===//
 
 import NIO
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+#else
 import CNIOBoringSSL
 import CNIOBoringSSLShims
+#endif
 
 // This is a neat trick. Swift lazily initializes module-globals based on when they're first
 // used. This lets us defer BoringSSL intialization as late as possible and only do it if people

--- a/Sources/NIOSSL/SSLErrors.swift
+++ b/Sources/NIOSSL/SSLErrors.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 /// Wraps a single error from BoringSSL.
 public struct BoringSSLInternalError: Equatable, CustomStringConvertible {

--- a/Sources/NIOSSL/SSLInit.swift
+++ b/Sources/NIOSSL/SSLInit.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 /// Initialize BoringSSL. Note that this function IS NOT THREAD SAFE, and so must be called inside
 /// either an explicit or implicit dispatch_once.

--- a/Sources/NIOSSL/SSLPKCS12Bundle.swift
+++ b/Sources/NIOSSL/SSLPKCS12Bundle.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 
 

--- a/Sources/NIOSSL/SSLPrivateKey.swift
+++ b/Sources/NIOSSL/SSLPrivateKey.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 /// An `NIOSSLPassphraseCallback` is a callback that will be invoked by NIOSSL when it needs to
 /// get access to a private key that is stored in encrypted form.
@@ -112,10 +116,14 @@ func globalBoringSSLPassphraseCallback(buf: UnsafeMutablePointer<CChar>?,
 /// to obtain an in-memory representation of a key from a buffer of
 /// bytes or from a file path.
 public class NIOSSLPrivateKey {
-    internal let ref: UnsafeMutablePointer<EVP_PKEY>
+    internal let _ref: UnsafeMutableRawPointer /*<EVP_PKEY>*/
+
+    internal var ref: UnsafeMutablePointer<EVP_PKEY> {
+        return self._ref.assumingMemoryBound(to: EVP_PKEY.self)
+    }
 
     private init(withReference ref: UnsafeMutablePointer<EVP_PKEY>) {
-        self.ref = ref
+        self._ref = UnsafeMutableRawPointer(ref) // erasing the type for @_implementationOnly import CNIOBoringSSL
     }
 
     /// A delegating initializer for `init(file:format:passphraseCallback)` and `init(file:format:)`.

--- a/Sources/NIOSSL/SSLPublicKey.swift
+++ b/Sources/NIOSSL/SSLPublicKey.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 /// An `NIOSSLPublicKey` is an abstract handle to a public key owned by BoringSSL.
 ///
@@ -21,10 +25,14 @@ import CNIOBoringSSL
 /// `NIOSSLCertificate` objects to be serialized, so that they can be passed to
 /// general-purpose cryptography libraries.
 public class NIOSSLPublicKey {
-    private let ref: UnsafeMutablePointer<EVP_PKEY>
+    private let _ref: UnsafeMutableRawPointer /*<EVP_PKEY>*/
+
+    private var ref: UnsafeMutablePointer<EVP_PKEY> {
+        return self._ref.assumingMemoryBound(to: EVP_PKEY.self)
+    }
 
     fileprivate init(withOwnedReference ref: UnsafeMutablePointer<EVP_PKEY>) {
-        self.ref = ref
+        self._ref = UnsafeMutableRawPointer(ref) // erasing the type for @_implementationOnly import CNIOBoringSSL
     }
 
     deinit {

--- a/Sources/NIOSSL/SecurityFrameworkCertificateVerification.swift
+++ b/Sources/NIOSSL/SecurityFrameworkCertificateVerification.swift
@@ -11,7 +11,12 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 //===----------------------------------------------------------------------===//
+
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 
 /// The current state of the platform verification helper, if one is in use.
 ///

--- a/Sources/NIOSSL/TLSConfiguration.swift
+++ b/Sources/NIOSSL/TLSConfiguration.swift
@@ -12,7 +12,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 
 /// Known and supported TLS versions.

--- a/Tests/NIOSSLTests/ByteBufferBIOTest.swift
+++ b/Tests/NIOSSLTests/ByteBufferBIOTest.swift
@@ -14,7 +14,11 @@
 
 import XCTest
 import NIO
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 @testable import NIOSSL
 
 

--- a/Tests/NIOSSLTests/NIOSSLALPNTest.swift
+++ b/Tests/NIOSSLTests/NIOSSLALPNTest.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import XCTest
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 import NIOTLS
 import NIOSSL

--- a/Tests/NIOSSLTests/NIOSSLIntegrationTest.swift
+++ b/Tests/NIOSSLTests/NIOSSLIntegrationTest.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import XCTest
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 @testable import NIOSSL
 import NIOTLS

--- a/Tests/NIOSSLTests/NIOSSLTestHelpers.swift
+++ b/Tests/NIOSSLTests/NIOSSLTestHelpers.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import Foundation
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 @testable import NIOSSL
 
 let samplePemCert = """

--- a/Tests/NIOSSLTests/TLSConfigurationTest.swift
+++ b/Tests/NIOSSLTests/TLSConfigurationTest.swift
@@ -13,7 +13,11 @@
 //===----------------------------------------------------------------------===//
 
 import XCTest
+#if compiler(>=5.1) && compiler(<5.2)
+@_implementationOnly import CNIOBoringSSL
+#else
 import CNIOBoringSSL
+#endif
 import NIO
 @testable import NIOSSL
 import NIOTLS

